compiler/rustc_const_eval/src/util/caller_location.rs

```diff
@@ -16,13 +16,19 @@ fn alloc_caller_location<'tcx>(
     line: u32,
     col: u32,
 ) -> MPlaceTy<'tcx> {
+    // Ensure that the filename itself does not contain nul bytes.
+    // This isn't possible via POSIX or Windows, but we should ensure no one
+    // ever does such a thing.
+    assert!(!filename.as_str().as_bytes().contains(&0));
+
     let loc_details = ecx.tcx.sess.opts.unstable_opts.location_detail;
     // This can fail if rustc runs out of memory right here. Trying to emit an error would be
     // pointless, since that would require allocating more memory than these short strings.
     let file = if loc_details.file {
-        ecx.allocate_str_dedup(filename.as_str()).unwrap()
+        let filename_with_nul = filename.as_str().to_owned() + "\0";
+        ecx.allocate_byte_slice_dedup(filename_with_nul.as_bytes()).unwrap()
     } else {
-        ecx.allocate_str_dedup("<redacted>").unwrap()
+        ecx.allocate_byte_slice_dedup("<redacted>\0".as_bytes()).unwrap()
     };
```

Comment:
I don't think this needs to be a place, so we can probably avoid those single-use APIs with something like
```rust
let filename = if loc_details.file { filename.as_str() } else { "<redacted>" };
let filename_with_nul = filename.to_owned() + "\0";
let ptr = ecx.allocate_bytes_dedup(filename_with_nul.as_bytes()).unwrap();
let wide_ptr = Immediate::new_slice(ptr.into(), filename_with_nul.len().try_into().unwrap());

// ...

ecx.write_immediate(file, &ecx.project_field(&location, 0).unwrap()).expect(...);
```
This loses the `map_provenance` but that's not load-bearing anyway.

Reply:
Done.